### PR TITLE
fix(statusChatInput): update mentions behavior when text is pasted

### LINKED ISSUE #4867: fix(chat/mentions) update mentions behavior when text is copied and pasted in chat text input
### Description
With https://github.com/status-im/status-desktop/pull/4857 if there is a mention in the text input it is not possible to have the cursor inside a mention, as well as when pointing to a mention with the mouse, this is treated as a link. However this is not the case when text that contains a mention is copied and then pasted to the text input.

1) the positions of the mentions in the clipboard text should be somehow stored so that the cursor jumps over the mention
2) when text is pasted and due to it being a rich text containing a link, the text is decorated to be underlined by default, this underline should be removed (try something like `var plainText = messageInputField.getFormattedText(0, messageInputField.length).replace(/text-decoration: underline;/g, " ") .replace(/span style=\"/g, "span style=\"text-decoration:none; ");`

### PR DESCRIPTION
**Pull request by alexandraB99**
Closes #4867 

Depends on https://github.com/status-im/dotherside/pull/59

### What does the PR do
Updated chat text input to store all mentions currently in the field and not allow the cursor to get into these when going either left right, or up/down. Also, fixed other copy/paste text issues, when for example a text is copied and pasted somewhere inline the existing string.

### Affected areas
chat text input

### Screenshot of functionality
https://user-images.githubusercontent.com/31625338/155813358-bffe1b34-04c6-4912-85e3-f4b8ba8fd9ec.mov

### CODE REVIEW

File: ui/imports/shared/status/StatusChatInput.qml [hunk 1]
```diff
@@ -383,17 +373,44 @@ Rectangle {
...
+                        var subStr = name.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
+                        var regex = new RegExp(subStr, 'gi'), result, indices = [];
+                        while ((result = regex.exec(copiedTextPlain))) {
+                            mentionsPos.push({"leftIndex": (result.index + copyTextStart - 1), "rightIndex": (result.index + copyTextStart + name.length)});
+                        }
+                    }
+                }
+            }
+            messageInputField.remove(copyTextStart, (copyTextStart + copiedTextPlain.length));
+            insertInTextInput(copyTextStart, copiedTextFormatted);
```
Comment (jrainville) [inline ui/imports/shared/status/StatusChatInput.qml:412]: One question/concern I have with this is , what happens if you copy text from the input, then copy text from outside the app (from your browser for example? I think the paste would then put the first copied text, which would confuse the user.

I'm not sure how we can fix that to be honest. Maybe it's a rare occurrence, or also maybe I misunderstood the code and it wouldn't do that anyway?

File: ui/imports/shared/status/StatusChatInput.qml [hunk 2]
```diff
@@ -383,17 +373,44 @@ Rectangle {
...
         // we can only get it in the `released` event
+
         if (paste) {
+            if (copiedTextPlain.includes("@")) {
+                copiedTextFormatted = copiedTextFormatted.replace(/underline/g, "none").replace(/span style="/g, "span style=\" text-decoration:none;");
+                for (var j = 0; j < dummyContactList.count; j++) {
+                    var name = dummyContactList.itemAtIndex(j).contactName;
+                    if (copiedTextPlain.indexOf(name) > -1) {
+                        name += '';
+                        copiedTextPlain += '';
```
Comment (jrainville) [inline ui/imports/shared/status/StatusChatInput.qml:402]: These two lines don't seem to do much. Are they to convert the variables to string?

**Review by jrainville — COMMENTED** [review]
Very nice work. The new behaviour is very clean

**Comment by noeliaSD** [comment]
Great job! The behavior is how I'd expect it has.

**Review by noeliaSD — APPROVED** [review]

File: ui/imports/shared/status/StatusChatInput.qml (hunk 1 above)
Comment (noeliaSD) [reply ui/imports/shared/status/StatusChatInput.qml:412]: Yeah! When you copy outside the app, it makes a mix of the text outside and the last copied text inside, writing as a firstly the inside one and then concatenate with the outside.

See the video for better understanding!

https://user-images.githubusercontent.com/97019400/156158592-2fa423a9-143d-41e9-88c2-de769e176bed.mov

**Review by noeliaSD — CHANGES_REQUESTED** [review]
Great job!! It's how I would expect it to work. 

Just only 2 few more things:

1. When sending the message that contains a mention.. Should it show the chat key of the contact or the mention itself? Now I can see that it shows the chat key.
2. When we position the cursor after a mention and go back with the left arrow button, if the mention is not in the first position of the chat input, it works fine, but if it is the first one, the user can write characters  within the text highlighted. See the attached video for a better understanding:


https://user-images.githubusercontent.com/97019400/156160001-e6aa70fd-d8b4-4399-8b23-f2af0c5e3783.mov

File: ui/imports/shared/status/StatusChatInput.qml (hunk 2 above)
Comment (alexandraB99) [reply ui/imports/shared/status/StatusChatInput.qml]: Removed

File: ui/imports/shared/status/StatusChatInput.qml (hunk 1 above)
Comment (alexandraB99) [reply ui/imports/shared/status/StatusChatInput.qml:417]: This should be fixed now, please check latest version, have in mind it depends on https://github.com/status-im/dotherside/pull/59 so you might need to remove DOtherSide lib and build again to see the changes.

**Comment by alexandraB99** [comment]
> Great job!! It's how I would expect it to work.
> 
> Just only 2 few more things:
> 
> 1. When sending the message that contains a mention.. Should it show the chat key of the contact or the mention itself? Now I can see that it shows the chat key.
> 2. When we position the cursor after a mention and go back with the left arrow button, if the mention is not in the first position of the chat input, it works fine, but if it is the first one, the user can write characters  within the text highlighted. See the attached video for a better understanding:
> 
>  Screen.Recording.2022-03-01.at.12.13.15.mov

@noeliaSD thanks for that. The first one I cannot reproduce and the second is happening before in master as well, it's a general issue related to formatting, not necessarily to pasting text. I looked at it a bit but it doesn't seem as an easy fix, would it be possible to file another bug for this?

**Comment by noeliaSD** [comment]
> > Great job!! It's how I would expect it to work.
> > Just only 2 few more things:
> > 
> > 1. When sending the message that contains a mention.. Should it show the chat key of the contact or the mention itself? Now I can see that it shows the chat key.
> > 2. When we position the cursor after a mention and go back with the left arrow button, if the mention is not in the first position of the chat input, it works fine, but if it is the first one, the user can write characters  within the text highlighted. See the attached video for a better understanding:
> > 
> > Screen.Recording.2022-03-01.at.12.13.15.mov
> 
> @noeliaSD thanks for that. The first one I cannot reproduce and the second is happening in master as well, it's a general issue related to formatting, not necessarily to pasting text. I looked at it a bit but it doesn't seem as an easy fix, would it be possible to file another bug for this?

Ok! It sound like a good idea to separate it in another issue if it seems more complex.

And regarding mentions shown as chat key when sent.. I can still reproduce it. I do nothing special (see the video). Maybe @jrainville can also test it. Maybe it can be, as well, managed in another separate issue..

https://user-images.githubusercontent.com/97019400/156314708-9f05655e-0b45-4dcd-b7b3-8c2e48660c27.mov

**Comment by jrainville** [comment]
The pubkey mention seems to happen when you mention yourself @noeliaSD . It was already logged as an issue. Either we'll disable mentionning yourself or it will be fixed with that issue.

**Review by jrainville — APPROVED** [review]
Great work

**Comment by alexandraB99** [comment]
Rebased, @noeliaSD could you resolve your requests if are not still valid please?

**Comment by noeliaSD** [comment]
> The pubkey mention seems to happen when you mention yourself @noeliaSD . It was already logged as an issue. Either we'll disable mentionning yourself or it will be fixed with that issue.

Oh! Nice to know! Thanks for the clarification!

**Review by noeliaSD — APPROVED** [review]

**Comment by abhi-status** [comment]
@alexandraB99 I can break the mention by hitting 'Ctrl+left key'

https://www.loom.com/share/2db64196682a48a49614947bacc9e397

**Review by iurimatias — APPROVED** [review]

**Pull request merged**